File: 1061-lexicographically-smallest-equivalent-string/1061-lexicographically-smallest-equivalent-string.ts
//we solve wirth union-find algo
//we initialize a map from 0 -25 to keep track of the alphabets a-z
//and we keep each element as the parent initially

function smallestEquivalentString(s1: string, s2: string, baseStr: string): string {
      const parent: Record<number, number> = {};
    for(let i = 0; i < 26; i++) parent[i] = i;  
    //remap all equivalent characters lexicographically in the parent hash_map
    for(let i = 0; i < s1.length; i++){
        let a = find(charCodeDiff(s1, i), parent);
        let b = find(charCodeDiff(s2, i), parent);
        if(a < b){
            parent[b] = a;
        } else if ( b < a){
            parent[a] = b;
        }
    }
    
    let ans = "";
    for(let i = 0; i < baseStr.length; i++){
        ans += String.fromCharCode(find(charCodeDiff(baseStr, i), parent) + 97);
    }
    return ans;
};

const charCodeDiff = (a: string, i: number) => {
    return a.charCodeAt(i) - "a".charCodeAt(0);
}

//recursively find the parent of a word
const find = (word, parent) => {
   return parent[word] === word ? word : find(parent[word], parent)
}

File: 1061-lexicographically-smallest-equivalent-string/1061-lexicographically-smallest-equivalent-string.py
class Solution:
    def __init__(self):
        self.parent = [i for i in range(26)]
        
    def smallestEquivalentString(self, s1: str, s2: str, baseStr: str) -> str:
        for i in range(len(s1)):
            a = self.find(ord(s1[i]) - 97)
            b = self.find(ord(s2[i]) - 97)
            if a > b:
                self.parent[a] = b
            elif b > a:
                self.parent[b] = a
        ans = ""
        for i in range(len(baseStr)):
            ans += chr(self.find(ord(baseStr[i]) - 97) + 97)
        return ans
    
    
    def find(self, word):
        if self.parent[word] == word:
            return word
        else:
            return self.find(self.parent[word])
        